File: protocols/gossip_protocol/GossipNode.js
/**
 * GossipNode - Epidemic broadcast protocol for distributed state sync
 */

const PeerManager = require('./PeerManager');
const MessageRouter = require('./MessageRouter');
const VectorClock = require('./VectorClock');

class GossipNode {
    constructor(config = {}) {
        this.config = {
            nodeId: config.nodeId || this.generateNodeId(),
            gossipInterval: config.gossipInterval || 5000,
            fanout: config.fanout || 3, // Number of peers to gossip to
            maxMessageAge: config.maxMessageAge || 300000, // 5 minutes
            ...config
        };

        this.peers = new PeerManager();
        this.router = new MessageRouter();
        this.vectorClock = new VectorClock(this.config.nodeId);
        
        this.messageStore = new Map(); // Seen messages with timestamps
        this.handlers = new Map();
        
        this.gossipTimer = null;
        this.isRunning = false;
    }

    async initialize() {
        // Setup message handlers
        this.setupDefaultHandlers();
        
        // Initialize transport (WebRTC, WebSocket, etc.)
        await this.initializeTransport();
    }

    setupDefaultHandlers() {
        // Billing sync handler
        this.registerHandler('billing_sync', require('./handlers/billing_sync'));
        
        // Auth broadcast handler
        this.registerHandler('auth_broadcast', require('./handlers/auth_broadcast'));
        
        // Presence heartbeat
        this.registerHandler('presence_heartbeat', require('./handlers/presence_heartbeat'));
    }

    registerHandler(type, handler) {
        this.handlers.set(type, handler);
        handler.initialize?.(this);
    }

    async initializeTransport() {
        // Initialize based on environment
        if (typeof RTCPeerConnection !== 'undefined') {
            this.transport = new WebRTCTransport(this.config);
        } else {
            this.transport = new WebSocketTransport(this.config);
        }
        
        await this.transport.initialize();
        
        this.transport.on('message', (msg) => this.handleIncomingMessage(msg));
        this.transport.on('peer_connected', (peer) => this.peers.addPeer(peer));
        this.transport.on('peer_disconnected', (peer) => this.peers.removePeer(peer));
    }

    start() {
        if (this.isRunning) return;
        
        this.isRunning = true;
        this.gossipTimer = setInterval(() => this.gossipRound(), this.config.gossipInterval);
        
        // Send initial presence
        this.broadcast('presence_heartbeat', {
            nodeId: this.config.nodeId,
            timestamp: Date.now(),
            capabilities: this.getCapabilities()
        });
    }

    stop() {
        this.isRunning = false;
        clearInterval(this.gossipTimer);
        this.transport?.close();
    }

    async gossipRound() {
        if (this.peers.size === 0) return;

        // Select random peers to gossip with
        const targets = this.peers.selectRandomPeers(this.config.fanout);
        
        // Get recent messages to share
        const messages = this.getRecentMessages();

        for (const peer of targets) {
            try {
                await this.transport.send(peer.id, {
                    type: 'gossip',
                    sender: this.config.nodeId,
                    vectorClock: this.vectorClock.get(),
                    messages: this.selectMessagesForPeer(peer, messages)
                });
            } catch (error) {
                this.peers.markPeerFailed(peer.id);
            }
        }
    }

    getRecentMessages() {
        const cutoff = Date.now() - this.config.maxMessageAge;
        const recent = [];
        
        for (const [id, msg] of this.messageStore) {
            if (msg.timestamp > cutoff) {
                recent.push(msg);
            } else {
                this.messageStore.delete(id); // Cleanup old messages
            }
        }
        
        return recent.sort((a, b) => b.timestamp - a.timestamp).slice(0, 50);
    }

    selectMessagesForPeer(peer, messages) {
        // Filter out messages peer likely already has
        return messages.filter(msg => {
            const peerVector = peer.lastVector || {};
            return !this.vectorClock.isDescendant(peerVector, msg.vectorClock);
        });
    }

    async handleIncomingMessage(message) {
        // Update vector clock
        this.vectorClock.update(message.vectorClock);

        // Process gossip messages
        if (message.type === 'gossip' && message.messages) {
            for (const msg of message.messages) {
                await this.processMessage(msg);
            }
        }

        // Direct message handling
        if (message.type !== 'gossip') {
            await this.processMessage(message);
        }
    }

    async processMessage(message) {
        const messageId = this.generateMessageId(message);
        
        // Check if already seen
        if (this.messageStore.has(messageId)) {
            return; // Duplicate, ignore
        }

        // Store message
        this.messageStore.set(messageId, {
            ...message,
            receivedAt: Date.now(),
            hops: (message.hops || 0) + 1
        });

        // Route to handler
        const handler = this.handlers.get(message.type);
        if (handler) {
            try {
                await handler.handle(message, this);
            } catch (error) {
                console.error(`[GossipNode] Handler error for ${message.type}:`, error);
            }
        }

        // Emit to local subscribers
        this.emit('gossip:message', message);

        // Forward if hops allow
        if (message.maxHops && message.hops < message.maxHops) {
            this.forwardMessage(message);
        }
    }

    forwardMessage(message) {
        // Epidemic propagation - gossip to random peers
        const targets = this.peers.selectRandomPeers(Math.ceil(this.config.fanout / 2));
        
        for (const peer of targets) {
            this.transport.send(peer.id, message).catch(() => {
                this.peers.markPeerFailed(peer.id);
            });
        }
    }

    broadcast(type, payload, options = {}) {
        const message = {
            type,
            payload,
            sender: this.config.nodeId,
            vectorClock: this.vectorClock.increment(),
            timestamp: Date.now(),
            hops: 0,
            maxHops: options.maxHops || 5,
            ...options
        };

        // Store locally
        const messageId = this.generateMessageId(message);
        this.messageStore.set(messageId, message);

        // Immediate broadcast to all peers
        for (const peer of this.peers.getAll()) {
            this.transport.send(peer.id, message).catch(() => {
                this.peers.markPeerFailed(peer.id);
            });
        }

        return messageId;
    }

    sendToPeer(peerId, type, payload) {
        return this.transport.send(peerId, {
            type,
            payload,
            sender: this.config.nodeId,
            timestamp: Date.now()
        });
    }

    generateNodeId() {
        return `gossip-${Math.random().toString(36).substr(2, 9)}`;
    }

    generateMessageId(message) {
        return `${message.sender}-${message.timestamp}-${message.type}`;
    }

    getCapabilities() {
        return {
            billing: true,
            auth: true,
            relay: true,
            mesh: true
        };
    }

    getStats() {
        return {
            nodeId: this.config.nodeId,
            peers: this.peers.size,
            messagesStored: this.messageStore.size,
            vectorClock: this.vectorClock.get()
        };
    }

    emit(event, data) {
        // Event emission for external listeners
    }
}

export default GossipNode;

File: custom-plugins/cordova-plugin-aicore/cordova-plugin-wifi-billing-agent/src/android/protocols/GossipNode.java
package com.wifibilling.agent.protocols;

import android.content.Context;
import android.net.wifi.p2p.WifiP2pConfig;
import android.net.wifi.p2p.WifiP2pDevice;
import android.net.wifi.p2p.WifiP2pManager;
import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class GossipNode {

    private Context context;
    private String nodeId;
    private int gossipInterval;
    private int fanout;
    private Handler handler;
    private ExecutorService executor;
    private Random random;

    private Map<String, Peer> peers = new ConcurrentHashMap<>();
    private Map<String, GossipMessage> messageStore = new ConcurrentHashMap<>();
    private Map<String, VectorClock> vectorClocks = new HashMap<>();

    private boolean isRunning = false;
    private Runnable gossipRunnable;

    // Transport layers
    private WebRTCDataChannel webRTC;
    private BluetoothMesh bluetoothMesh;

    public GossipNode(Context context, JSONObject config) {
        this.context = context;
        this.nodeId = config.optString("agentId", "unknown");
        this.gossipInterval = config.optInt("gossipInterval", 5000);
        this.fanout = config.optInt("fanout", 3);
        this.handler = new Handler(Looper.getMainLooper());
        this.executor = Executors.newCachedThreadPool();
        this.random = new Random();

        // Initialize transports
        this.webRTC = new WebRTCDataChannel(context, this);
        this.bluetoothMesh = new BluetoothMesh(context, this);
    }

    public void start() {
        if (isRunning)
            return;

        isRunning = true;

        // Start transports
        webRTC.initialize();
        bluetoothMesh.initialize();

        // Start gossip rounds
        gossipRunnable = new Runnable() {
            @Override
            public void run() {
                if (isRunning) {
                    performGossipRound();
                    handler.postDelayed(this, gossipInterval);
                }
            }
        };
        handler.postDelayed(gossipRunnable, gossipInterval);
    }

    public void stop() {
        isRunning = false;
        handler.removeCallbacks(gossipRunnable);
        webRTC.shutdown();
        bluetoothMesh.shutdown();
        executor.shutdown();
    }

    private void performGossipRound() {
        if (peers.isEmpty())
            return;

        // Select random peers to gossip with
        Peer[] peerArray = peers.values().toArray(new Peer[0]);
        int targetCount = Math.min(fanout, peerArray.length);

        for (int i = 0; i < targetCount; i++) {
            Peer target = peerArray[random.nextInt(peerArray.length)];
            sendGossipToPeer(target);
        }
    }

    private void sendGossipToPeer(Peer peer) {
        executor.execute(() -> {
            try {
                // Get recent messages
                JSONArray messages = getRecentMessages();

                JSONObject gossip = new JSONObject();
                gossip.put("type", "gossip");
                gossip.put("sender", nodeId);
                gossip.put("vectorClock", getVectorClock().toJSON());
                gossip.put("messages", messages);

                // Send via appropriate transport
                if (peer.supportsWebRTC) {
                    webRTC.send(peer.id, gossip);
                } else {
                    bluetoothMesh.send(peer.id, gossip);
                }

            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }

    public void handleIncomingMessage(String peerId, JSONObject message) {
        try {
            // Update vector clock
            VectorClock receivedClock = VectorClock.fromJSON(message.getJSONObject("vectorClock"));
            mergeVectorClock(receivedClock);

            // Process messages
            if (message.has("messages")) {
                JSONArray messages = message.getJSONArray("messages");
                for (int i = 0; i < messages.length(); i++) {
                    processMessage(messages.getJSONObject(i));
                }
            }

            // Handle direct messages
            if (!message.optString("type", "").equals("gossip")) {
                processMessage(message);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void processMessage(JSONObject msg) throws JSONException {
        String messageId = msg.getString("id");

        // Check if already seen
        if (messageStore.containsKey(messageId)) {
            return;
        }

        // Store message
        GossipMessage gossipMsg = new GossipMessage();
        gossipMsg.id = messageId;
        gossipMsg.type = msg.getString("type");
        gossipMsg.payload = msg.getJSONObject("payload");
        gossipMsg.timestamp = msg.getLong("timestamp");
        gossipMsg.sender = msg.getString("sender");
        gossipMsg.vectorClock = VectorClock.fromJSON(msg.getJSONObject("vectorClock"));

        messageStore.put(messageId, gossipMsg);

        // Route to appropriate handler
        routeMessage(gossipMsg);
    }

    private void routeMessage(GossipMessage msg) {
        // Emit to core agent for routing to appropriate agent
        // billing_sync -> BillingAgent
        // auth_broadcast -> AuthAgent
        // presence_heartbeat -> update peer registry

        switch (msg.type) {
            case "billing_sync":
                // Forward to billing agent
                break;
            case "auth_broadcast":
                // Forward to auth agent
                break;
            case "presence_heartbeat":
                updatePeerPresence(msg);
                break;
        }
    }

    private void updatePeerPresence(GossipMessage msg) {
        Peer peer = peers.get(msg.sender);
        if (peer == null) {
            peer = new Peer();
            peer.id = msg.sender;
            peers.put(msg.sender, peer);
        }
        peer.lastSeen = System.currentTimeMillis();
        peer.capabilities = msg.payload.optJSONObject("capabilities");
    }

    public String broadcast(String type, JSONObject payload) {
        try {
            String messageId = generateMessageId();

            JSONObject message = new JSONObject();
            message.put("id", messageId);
            message.put("type", type);
            message.put("payload", payload);
            message.put("sender", nodeId);
            message.put("timestamp", System.currentTimeMillis());
            message.put("vectorClock", incrementVectorClock().toJSON());
            message.put("ttl", 10); // Time-to-live hops

            // Store locally
            GossipMessage gossipMsg = new GossipMessage();
            gossipMsg.id = messageId;
            gossipMsg.type = type;
            gossipMsg.payload = payload;
            gossipMsg.timestamp = System.currentTimeMillis();
            gossipMsg.sender = nodeId;
            messageStore.put(messageId, gossipMsg);

            // Broadcast to all peers
            for (Peer peer : peers.values()) {
                sendMessageToPeer(peer, message);
            }

            return messageId;

        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    private void sendMessageToPeer(Peer peer, JSONObject message) {
        executor.execute(() -> {
            try {
                if (peer.supportsWebRTC) {
                    webRTC.send(peer.id, message);
                } else {
                    bluetoothMesh.send(peer.id, message);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }

    public JSONObject getPeers() throws JSONException {
        JSONObject result = new JSONObject();
        JSONArray peerArray = new JSONArray();

        for (Peer peer : peers.values()) {
            JSONObject p = new JSONObject();
            p.put("id", peer.id);
            p.put("lastSeen", peer.lastSeen);
            p.put("latency", peer.latency);
            p.put("supportsWebRTC", peer.supportsWebRTC);
            p.put("supportsBluetooth", peer.supportsBluetooth);
            peerArray.put(p);
        }

        result.put("peers", peerArray);
        result.put("count", peers.size());
        return result;
    }

    public int getPeerCount() {
        return peers.size();
    }

    public JSONObject getStats() throws JSONException {
        JSONObject stats = new JSONObject();
        stats.put("nodeId", nodeId);
        stats.put("peers", peers.size());
        stats.put("messagesStored", messageStore.size());
        stats.put("running", isRunning);
        return stats;
    }

    private JSONArray getRecentMessages() throws JSONException {
        JSONArray recent = new JSONArray();
        long cutoff = System.currentTimeMillis() - 300000; // 5 minutes

        for (GossipMessage msg : messageStore.values()) {
            if (msg.timestamp > cutoff) {
                JSONObject m = new JSONObject();
                m.put("id", msg.id);
                m.put("type", msg.type);
                m.put("payload", msg.payload);
                m.put("timestamp", msg.timestamp);
                m.put("sender", msg.sender);
                m.put("vectorClock", msg.vectorClock.toJSON());
                recent.put(m);
            }
        }

        return recent;
    }

    private VectorClock getVectorClock() {
        return vectorClocks.getOrDefault(nodeId, new VectorClock(nodeId));
    }

    private VectorClock incrementVectorClock() {
        VectorClock clock = getVectorClock();
        clock.increment();
        vectorClocks.put(nodeId, clock);
        return clock;
    }

    private void mergeVectorClock(VectorClock other) {
        VectorClock local = getVectorClock();
        local.merge(other);
    }

    private String generateMessageId() {
        return nodeId + "-" + System.currentTimeMillis() + "-" + random.nextInt(10000);
    }

    // Peer class
    private static class Peer {
        String id;
        long lastSeen;
        long latency;
        boolean supportsWebRTC;
        boolean supportsBluetooth;
        JSONObject capabilities;
    }

    // Message class
    private static class GossipMessage {
        String id;
        String type;
        JSONObject payload;
        long timestamp;
        String sender;
        VectorClock vectorClock;
        int hops;
    }
}
